CRM/Contribute/Form/UpdateSubscription.php

```diff
@@ -116,12 +122,23 @@ public function preProcess() {
       CRM_Custom_Form_CustomData::setDefaultValues($this);
     }
 
-    $this->assign('editableScheduleFields', array_diff($this->editableScheduleFields, $alreadyHardCodedFields));
-
-    if ($this->_subscriptionDetails->contact_id) {
-      $contactDetails = CRM_Contact_BAO_Contact::getContactDetails($this->_subscriptionDetails->contact_id);
-      $this->_donorEmail = $contactDetails[1];
+    // amount, installments are added to the form as separate elements
+    // everything that is in the smarty variable editableScheduleFields will be added (again)
+    $editableScheduleFieldsToBeAddedToForm = [];
+    foreach ($this->editableScheduleFields as $field) {
+      if (in_array($field, ['amount', 'installments'])) {
+        continue;
+      }
+      $editableScheduleFieldsToBeAddedToForm[] = $field;
     }
+    $this->assign('editableScheduleFields', $editableScheduleFieldsToBeAddedToForm);
+
+    $email = \Civi\Api4\Email::get(FALSE)
+      ->addWhere('contact_id', '=', $this->contributionRecur['contact_id'])
+      ->addOrderBy('is_primary', 'DESC')
+      ->execute()
+      ->first();
+    $this->_donorEmail = $email['email'] ?? '';
```

Comment:
I suspect we don't need this property anymore but as long as `CRM_Contact_Form_ContactFormTrait` is used on the form we can use `$this->getContactValue('email_primary.email') here